server/src/testFixtures/java/io/crate/testing/SQLTransportExecutor.java

```diff
@@ -499,48 +497,80 @@ private BulkResponse executeBulk(String stmt, Object[][] bulkArgs, TimeValue tim
-            // been removed by the master so that the health check applies to the set of nodes we expect to be part of the cluster.
-            request.waitForNodes(Integer.toString(expectedNumNodes));
-        }
-        ClusterHealthResponse actionGet = FutureUtils.get(client.admin().cluster().health(request));
-        if (state == ClusterHealthStatus.YELLOW) {
-            assertThat(actionGet.getStatus()).satisfiesAnyOf(
-                s -> assertThat(s).isEqualTo(state),
-                s -> assertThat(s).isEqualTo(ClusterHealthStatus.GREEN));
-        } else {
-            assertThat(actionGet.getStatus()).isEqualTo(state);
-        }
-        return actionGet.getStatus();
+    @SuppressWarnings("unchecked")
+    private void ensureState(ClusterHealthStatus state, @Nullable Integer expectedNumNodes) throws Exception {
+        ESTestCase.assertBusy(() -> {
+            // First, check the global cluster health. If e.g. the cluster hasn't elected a master node, the 2nd query
+            // containing the `sys.allocations` table will fail this table is routed to the master node.
+            var clusterHealthResponse = exec("select health, description from sys.cluster_health");
+            String color = (String) clusterHealthResponse.rows()[0][0];
+            String description = (String) clusterHealthResponse.rows()[0][1];
+            var assertion = assertThat(color)
+                .withFailMessage("Expected cluster to be %s, but was %s, reason: %s", state.name(), color, description);
+            if (state == ClusterHealthStatus.GREEN) {
+                assertion.isEqualTo(state.name());
+            } else {
+                assertion.satisfiesAnyOf(
+                    x -> assertThat(x).isEqualTo("GREEN"),
+                    x -> assertThat(x).isEqualTo("YELLOW")
+                );
+            }
+            // If the cluster is green/yellow, we can also check the state of the nodes & shards
+            var response = exec(
+                """
+                select
+                    (select count(*) from sys.nodes) as num_nodes,
+                    (select health from sys.cluster_health) as color,
+                    (select
+                        {
+                            all_states = array_unique(array_agg(current_state)),
+                            primary_states = array_unique(array_agg(current_state) FILTER (WHERE primary = true)),
+                            relocating = count(*) FILTER (WHERE current_state = 'RELOCATING')
+                        } as shards_state
+                        FROM sys.allocations
+                    )
+                """
+            );
+            if (expectedNumNodes != null) {
+                assertThat(response.rows()[0][0]).isEqualTo((long) expectedNumNodes);
+            }
+            color = (String) response.rows()[0][1];
```

Comment:
why do we need to re-assert the color?

Reply:
To be sure nothing has changed in between the 2 SQL statements. Do you think this isn't needed?